packages/test-utils/src/reporter.ts

```diff
@@ -920,44 +917,83 @@ class ReplayReporter<TRecordingMetadata extends UnstructuredMetadata = Unstructu
 
       let uploadableResults = this._uploadableResults.get(result.testId);
       if (!uploadableResults) {
-        uploadableResults = [];
+        uploadableResults = {
+          executions: {},
+          uploadedStatuses: {
+            passed: false,
+            failed: false,
+          },
+        };
         this._uploadableResults.set(result.testId, uploadableResults);
       }
-      uploadableResults.push(result);
+      let executions = uploadableResults.executions[result.executionGroupId];
+      if (!executions) {
+        executions = [];
+        uploadableResults.executions[result.executionGroupId] = executions;
+      }
+      executions.push(result);
 
       if (result.result === "passed" || result.attempt >= result.maxAttempts) {
-        this._enqueueUploads(uploadableResults);
+        this._enqueueUploads(uploadableResults, result.executionGroupId);
       }
     }
   }
 
-  private _enqueueUploads(results: UploadableTestRunResult<TRecordingMetadata>[]) {
-    const latestResult = results[results.length - 1];
-    assert(!!latestResult, "Expected at least one result in the list");
-
-    let toUpload: typeof results | undefined;
+  private _enqueueUploads(
+    result: UploadableTestResult<TRecordingMetadata>,
+    executionGroupId: string
+  ) {
+    const executions = result.executions[executionGroupId];
+    const latestExecution = last(executions);
+    assert(!!latestExecution, "Expected at least one execution in the list");
 
+    let toUpload: typeof executions | undefined;
     switch (this._uploadStatusThreshold) {
       case "all":
-        toUpload = this._minimizeUploads ? [latestResult] : results;
+        // even when `minimizeUploads` is combined with `repeatEach` it always makes sense to upload the latest result
+        // otherwise, we could upload a single successful attempt without uploading a potential failure of the same test
+        // coming from a different `repeatEachIndex`
+        toUpload = this._minimizeUploads ? [latestExecution] : executions;
```

Comment:
This looks wrong to me.

What if the test was flaky? We're only getting the passing recording this way.

Reply:
I wonder if this should be rewritten so that we have the pass/fail/flaky check on the outside and then the minimize check within?

Reply:
Something like this?

```js
const status = computeStatus();
switch (status) {
  case "failed": {
    switch (this._uploadStatusThreshold) {
      case "all":
      case "failed":
      case "failed-and-flaky": {
        toUpload = this._minimizeUploads ? [latestExecution] : executions;
        break;
      }
    }
    break;
  }
  case "flaky": {
    switch (this._uploadStatusThreshold) {
      case "all":
      case "failed-and-flaky": {
        if (this._minimizeUploads) {
          toUpload = [executions[0], executions[executions.length - 1]];
        } else {
          toUpload = executions;
        }
        break;
      }
    }
    break;
  }
  case "pass": {
    switch (this._uploadStatusThreshold) {
      case "all": {
        toUpload = this._minimizeUploads ? [latestExecution] : executions;
        break;
      }
    }
    break;
  }
}
```

Reply:
This function uses executions from a single execution group (that currently is equivalent to `repeatEachIndex`) but the uploads are minimized per test id.

We'd have an easier time if we'd push this whole decision until `onEnd` (that runs after everything gets completed). I decided to keep it in `onTestEnd` to start uploads asap - so we only have to await those upload jobs in `onEnd`. Maybe it's not worth it but it felt nice to have it "paralellized".

We can't determine if a test case is flaky based on a signal execution group because both could have only single executions and each of them could be of a different status. I thought about having an aggregate status on this uploadable result but it felt like a complication at a time.

The version above misses that those flakes can be spread across execution groups. I like your suggestion to reorganize this, I'll try to come up with something better.

Reply:
I have rewritten it based on ur suggestion. I think it's much cleaner now but the end result isn't exactly the same as your snippet, could u take another look?